analysis/complex.lean

```diff
@@ -0,0 +1,334 @@
+begin
+intro H,rw [←eta z,←eta w,H.left,H.right],
+end
+
+-- simp version
+
+theorem eq_iff_re_eq_and_im_eq (z w : complex) : z=w ↔ z.re=w.re ∧ z.im=w.im :=
+begin
+split,
+  intro H,rw [H],split;trivial,
+exact eq_of_re_eq_and_im_eq _ _,
+end
+
+lemma proj_re (r0 i0 : real) : (complex.mk r0 i0).re = r0 := rfl
+lemma proj_im (r0 i0 : real) : (complex.mk r0 i0).im = i0 := rfl
+
+local attribute [simp] eq_iff_re_eq_and_im_eq proj_re proj_im
+
+def of_real : ℝ → ℂ := λ x, { re := x, im := 0 }
+
+instance coe_real_complex : has_coe ℝ ℂ := ⟨of_real⟩
+instance has_zero_complex : has_zero complex := ⟨of_real 0⟩
+instance has_one_complex : has_one complex := ⟨of_real 1⟩
+instance inhabited_complex : inhabited complex := ⟨0⟩
+
+def I : complex := {re := 0, im := 1}
+
+def conjugate (z : complex) : complex := {re := z.re, im := -(z.im)}
+
+-- Are these supposed to be protected too?
+
+protected def add : complex → complex → complex :=
+λ z w, { re :=z.re+w.re, im:=z.im+w.im}
+
+protected def neg : complex → complex :=
+λ z, { re := -z.re, im := -z.im}
+
+protected def mul : complex → complex → complex :=
+λ z w, { re := z.re*w.re - z.im*w.im,
+         im := z.re*w.im + z.im*w.re}
+
+def norm_squared : complex → real :=
+λ z, z.re*z.re+z.im*z.im
+
+protected def inv : complex → complex :=
+λ z,  { re := z.re / norm_squared z,
+        im := -z.im / norm_squared z }
+
+instance : has_add complex := ⟨complex.add⟩
+instance : has_neg complex := ⟨complex.neg⟩
+instance : has_mul complex := ⟨complex.mul⟩
+instance : has_inv complex := ⟨complex.inv⟩
+
+lemma proj_zero_re : (0:complex).re=0 := rfl
+lemma proj_zero_im : (0:complex).im=0 := rfl
+lemma proj_one_re : (1:complex).re=1 := rfl
+lemma proj_one_im : (1:complex).im=0 := rfl
+lemma proj_I_re : complex.I.re=0 := rfl
+lemma proj_I_im : complex.I.im=1 := rfl
+lemma proj_conj_re (z : complex) : (conjugate z).re = z.re := rfl
+lemma proj_conj_im (z : complex) : (conjugate z).im = -z.im := rfl
+lemma proj_add_re (z w: complex) : (z+w).re=z.re+w.re := rfl
+lemma proj_add_im (z w: complex) : (z+w).im=z.im+w.im := rfl
+lemma proj_add_re' (z w: complex) : (complex.add z w).re=z.re+w.re := rfl
+lemma proj_add_im' (z w: complex) : (complex.add z w).im=z.im+w.im := rfl
+lemma proj_add_re'' (z w: complex) : (has_add.add z w).re=z.re+w.re := rfl
+lemma proj_add_im'' (z w: complex) : (has_add.add z w).im=z.im+w.im := rfl
+lemma proj_neg_re (z: complex) : (-z).re=-z.re := rfl
+lemma proj_neg_im (z: complex) : (-z).im=-z.im := rfl
+lemma proj_neg_re' (z: complex) : (complex.neg z).re=-z.re := rfl
+lemma proj_neg_im' (z: complex) : (complex.neg z).im=-z.im := rfl
+
+local attribute [simp] proj_zero_re proj_zero_im proj_one_re proj_one_im
+local attribute [simp] proj_I_re proj_I_im proj_conj_re proj_conj_im
+local attribute [simp] proj_add_re proj_add_im proj_add_re' proj_add_im' proj_add_re'' proj_add_im''
+local attribute [simp] proj_neg_re proj_neg_im proj_neg_re' proj_neg_im'
+
+meta def crunch : tactic unit := do
```

Comment:
I don't think you need this method. Just replace `by crunch` with `by simp [add_mul, mul_add, eq_iff_re_eq_and_im_eq] {contextual:=tt}`